### PR TITLE
Add internal solution for Village (Minimum) (boi-20)

### PR DESCRIPTION
**Pull request by ChuangSheep**
_If any of the below don't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which include the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making the changes.

### CODE REVIEW

File: solutions/baltic-20-VillageMinimum.mdx [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+		adj[a].push_back(b);
+		adj[b].push_back(a);
+		changed_to[i] = i;
+	}
+	changed_to[N - 1] = N - 1;
+
+	int total_length = dfs(0, -1, changed_to, adj);
+
+	cout << total_length << endl;
+	for (auto &i : changed_to) {
```
Comment (SansPapyrus683) [inline solutions/baltic-20-VillageMinimum.mdx:99]: can you not use auto

File: solutions/baltic-20-VillageMinimum.mdx [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+/**
+ * Perform a depth-first search on the given node and swap villagers if
+ * necessary.
+ * @return the sum of length villagers need to travel
+ */
+int dfs(int node, int parent, vector<int> &changed_to,
+		const vector<vector<int>> &adj) {
+	int length_added = 0;
+
+	for (auto &child : adj[node]) {
```
Comment (SansPapyrus683) [inline solutions/baltic-20-VillageMinimum.mdx:50]: yeah here too

File: solutions/baltic-20-VillageMinimum.mdx [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+		if (child == parent) continue;
```
Comment (SansPapyrus683) [inline solutions/baltic-20-VillageMinimum.mdx:51]: braces

File: solutions/baltic-20-VillageMinimum.mdx [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	 * position with its parent node.
+	 */
+	if (parent >= 0 && changed_to[node] == node) {
+		changed_to[node] = changed_to[parent];
+		changed_to[parent] = node;
+		length_added += 2;
+	/*
+	 * If the current node is the root and did not swap its position yet, swap
+	 * it with any of its children.
+	 */
```
Comment (SansPapyrus683) [inline solutions/baltic-20-VillageMinimum.mdx:66]: i'd put this within the `else if` thing

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/baltic-20-VillageMinimum.mdx [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+depth-first search on the given tree. For each node (villager) that is still at
+its original position, we swap it with its only neighbor, namely its parent
+node. For the root node which does not have a parent node, we can swap it with
+any of its children.
+
+By swapping two nodes where at least one of them is not processed yet, we can
+guarantee that no node will be swapped back to its original position. We can
+also show that swapping between neighbors is the optimal solution: Suppose we
+would swap two unprocessed nodes which are not neighbors. The distance the two
+nodes need to travel would then be equal or larger than 4. If we instead just
```
Comment (SansPapyrus683) [inline solutions/baltic-20-VillageMinimum.mdx:22]: 4 seems like a really specific number, could you explain this w/ an example maybe?

Comment (SansPapyrus683) [reply solutions/baltic-20-VillageMinimum.mdx]: i'd add an illustration please

File: solutions/baltic-20-VillageMinimum.mdx [hunk 6]
```diff
@@ -19,12 +19,21 @@ By swapping two nodes where at least one of them is not processed yet, we can
...
 also show that swapping between neighbors is the optimal solution: Suppose we
 would swap two unprocessed nodes which are not neighbors. The distance the two
-nodes need to travel would then be equal or larger than 4 (they are travelling
-over at least one transiting node between them). If we instead just swap them
+nodes need to travel would then be equal or larger than 4. If we instead just swap them
 with their respective neighbors, the distance they need to travel is exactly 4.
 Therefore, swapping nodes that are not neighbors never lead to a better
 solution.
 
+To illustrate this idea, let's consider the following example: Nodes 2 and 3 are not swapped yet. Now,
```
Comment (SansPapyrus683) [inline solutions/baltic-20-VillageMinimum.mdx]: i'd put the illustration right below the first sentence

File: solutions/baltic-20-VillageMinimum.mdx [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+nodes need to travel would then be equal or larger than 4. If we instead just swap them
+with their respective neighbors, the distance they need to travel is exactly 4.
+Therefore, swapping nodes that are not neighbors never lead to a better
+solution.
+
+To illustrate this idea, let's consider the following example: 
+
+![Example 1](baltic-20-VillageMinimum/example1.png)
+
+Nodes 2 and 3 are not swapped yet. Now, if we swap node 3 and 4 directly, the
```
Comment (SansPapyrus683) [inline solutions/baltic-20-VillageMinimum.mdx:31]: i don't see a node 4 in the diagram?

Comment (ChuangSheep) [reply solutions/baltic-20-VillageMinimum.mdx:31]: opps gonna fix that
<del>changed the diagramm in between and forgot to update the text</del>

**Review by SansPapyrus683 — APPROVED** [review]
yeah!!

**Comment by ChuangSheep** [comment]
@ryanchou-dev pls :blobheart:

**Review by ryanchou-dev — APPROVED** [review]
looks good!

**Pull request merged**